### PR TITLE
Don't execute completions automatically if head does not exist

### LINKED ISSUE #2365: Don't complete arguments for non-existent commands
Hi,

I use fish 2.2.0 on Arch Linux and if I open a new shell (the bug does not seem to occur twice in the same session) and type `xrandr` + space, I get the following output:

```
$ xrandr xrandr may be found in the following packages:
  extra/xorg-xrandr 1.4.3-1     /usr/bin/xrandr
/usr/share/fish/completions/xrandr.fish (line 1): xrandr -v | grep RandR | sed 's/^.\+\s\([0-9\.]\+\)$/\1/'
                                                  ^
in command substitution
        called on line -1 of file /usr/share/fish/completions/xrandr.fish

from sourcing file /usr/share/fish/completions/xrandr.fish
        called on standard input

in command substitution
        called on standard input

expr: syntax error
```

### PR DESCRIPTION
**Pull request by mqudsi**
This fixes the issues with fish automatically spewing error output to the
console caused by autoloaded completions attempting to generate results
to provide the autosuggestion thread with data.

What we do is prevent completions from being considered as an
autosuggestion source if the command being typed out is not a
fully-formed head that exists. This only addresses *automatically*
generated errors, such as that when typing in `zfs` or `git` at the
prompt on an installation that does not have these utilities. This does
_not_ prevent errors caused by typing in `zfs <TAB>` or `git <TAB>` when
these tools are not installed.

Closes #2365 and possibly others; it has many duplicates such as #4688.

### CODE REVIEW

**Comment by mqudsi** [comment]
My only question here is whether my assumption that it's ok to ignore completions as an autosuggestion source if the command does not exist is a sane one.

File: src/reader.cpp [hunk 1]
```diff
@@ -1173,9 +1183,24 @@ static std::function<autosuggestion_result_t(void)> get_autosuggestion_performer
...
+
+        // Try normal completions last
+        // But not unless the head has been fully typed out and exists/is valid.
+        const wchar_t *first_space = wcschr(first_char, L' ');
+        if (first_space == nullptr) {
+            return nothing; //the head has not been typed out yet
+        }
+        wcstring head(first_char, first_space);
+
+        // Ideally, we should use function_exists and not function_exists_no_autoload here,
```
Comment (faho) [inline src/reader.cpp]: Why? The "no_autoload" version can still tell you if the function exists (or rather is _supposed_ to exist - i.e. a file that should contain a function of that name does).

**Comment by faho** [comment]
>it has many duplicates such as #4688.

That's not a duplicate - did you mean a different issue?

----

This seems nice, and is also the much more pressing issue, compared to #4783.

File: src/reader.cpp (hunk 1 above)
Comment (mqudsi) [reply src/reader.cpp]: I didn't realize that. Awesome.

**Comment by mqudsi** [comment]
I meant a different issue, copy-and-paste fail! Not worth looking up.

**Pull request closed without merging**